src/inc/til/spsc.h

```diff
@@ -0,0 +1,526 @@
+                // The following detects whether we'd be wrapping around the end of the ring buffer
+                // and splits the range into the first half [mine, _capacity).
+                // If acquire() is called again it'll return [0, theirs).
+                end = end > begin ? end : _capacity;
+
+                // Of course we also need to ensure to not return more than we've been asked for.
+                end = std::min(end, begin + slots);
+
+                // "next" will contain the value that's stored into "mine" when release() is called.
+                // It's basically the same as "end", but with the revolution flag spliced in.
+                // If we acquired the range [mine, _capacity) "end" will equal _capacity
+                // and thus wrap around the ring buffer. The next value for "mine" is thus the
+                // position zero | the flipped "revolution" (and 0 | x == x).
+                auto revolution = myPos & revolution_flag;
+                auto next = end != _capacity ? end | revolution : revolution ^ revolution_flag;
+
+                return {
+                    begin,
+                    end,
+                    next,
+                };
+            }
+
+            void release(atomic_size_type& mine, acquisition acquisition) noexcept
+            {
+                mine.store(acquisition.next, std::memory_order_release);
+                mine.notify_one();
+            }
+
+            T* const _data;
+            const size_type _capacity;
+
+            std::atomic<bool> _eitherSideDropped{ false };
+
+            atomic_size_type _producer;
+            atomic_size_type _consumer;
+        };
+    }
+
+    template<typename T>
+    struct sender
+    {
+        explicit sender(details::arc<T>* arc) noexcept :
+            _arc(arc) {}
+
+        sender<T>(const sender<T>&) = delete;
+        sender<T>& operator=(const sender<T>&) = delete;
+
+        sender(sender<T>&& other) noexcept
+        {
+            drop();
+            _arc = std::exchange(other._arc, nullptr);
+        }
+
+        sender<T>& operator=(sender<T>&& other) noexcept
+        {
+            drop();
+            _arc = std::exchange(other._arc, nullptr);
+        }
+
+        ~sender()
+        {
+            drop();
+        }
+
+        // emplace constructs an item in-place at the end of the queue.
+        // It returns true, if the item was successfully placed within the queue.
+        // The return value will be false, if the receiver is gone.
+        template<typename... Args>
+        bool emplace(Args&&... args) const
+        {
+            auto acquisition = _arc->producer_acquire(1);
+            if (!acquisition)
+            {
+                return false;
+            }
+
+            auto data = _arc->data();
+            auto begin = data + acquisition.begin;
+            new (begin) T(std::forward<Args>(args)...);

```

Comment:
If `producer_release()` isn't called the state in the `details::arc` won't be changed, and the next call to either method will yield the same `acquisition` value again. But this in turn also means we need to ensure that we call `producer_release()` every time the buffer contents actually were changed.

Luckily in the case of these two methods we're fine because `new` will destroy any partially constructed values and [`std::uninitialized_move_n`](https://en.cppreference.com/w/cpp/memory/uninitialized_move_n) will destroy the range of values that has been moved up to the point of the exception. In fact I couldn't even call `producer_release()` if I wanted to, because no valid data has been written to the queue's buffer if an exception happens during construction.

The other methods should be fine as well. `receiver<T>::pop_n` for instance might fail when moving the items, but if it does the range of data in the queue's buffer is still technically valid afterwards, which is why `consumer_release()` doesn't need to be called. If you call `pop_n` again it'll then yield empty, but valid object instances.

So as far as I can see the only method I could improve in that regard are `sender<T>::move_n` and `receiver<T>::pop_n`. But on the other hand I thought, that this level correctness would be overkill for this PR. So considering the "technical" correctness of this PR as it is, I was contempt with the simple calls to `std::uninitialized_move_n` and `std::move` respectively. What do you think about this?

Reply:
Oh yeah I didn't think about this. I guess it's fine if it would be overkill.